### DISCUSSION TITLE
Cron jobs with docker-compose

### DISCUSSION THREAD

**Discussion by Thulkash** [firefly-iii/firefly-iii#7731] [Firefly III] [answered]
Hello,
First of all, thanks JC5 for your amazing project.

I deployed Firefly III with docker-compose and it's working great.

I'm looking for triggering "future" cron jobs, as you describe it at the end of your documentation https://docs.firefly-iii.org/firefly-iii/advanced-installation/cron/ , with my docker-compose environment. How could I do it please?

If it is not possible, could you create a field (in the page Reccuring transactions for example) with a number of days (0 by default) that will be added to the present day?

And in the same documentation, you precise that you need to the token found in our /profile but it seems that the correct one is from the .env file (STATIC_CRON_TOKEN), am I right? This is pretty confusing.

Regarding the STATIC_CRON_TOKEN value in your .env.sample, you precise that "You can set this variable from a file by appending it with _FILE". What is the point? Because we need to write the token in the YAML file.

Thanks for your answers

**Comment by JC5** [comment]
> How could I do it please?

"docker exec -it (container)" and then the command.

**Reply by Thulkash** [reply]
Thanks, but as you said, it's not possible in the firefly docker to execute cron jobs, and I'd like to automate the command with docker-compose in the cron part.

If we can't automate this, may I suggest to add a number field, in the page Reccuring transactions, that will be added to the current date? But I really don't know how it works ^^

**Reply by JC5** [reply]
Ehhh, it's totally possible? Did I misunderstand the question?

**Reply by JC5** [reply]
Ah i think I know what you mean. You want to automate that it always does $date + x days, right?

**Reply by Thulkash** [reply]
Yes, exactly.

I need this because my VISA cards are only debited the first week day of the month "M+1" (delayed debit card) and I refill my account each end of the month "M" like this:
         VISA amount of month "M" (the current one)
      + all reccuring transactions of month "M+1"
      + difference to only have 300€ left on my bank account.

I'm using Grisbi since few years for my personal finances. I set its parameters to execute all reccuring transactions of "current day + 25 days".
Thus, each end of the month M, I know how much money I need to transfer on my bank account to cover the month "M+1" (+ 300€ left on my account, just in case)

**Reply by JC5** [reply]
Ah sorry. That's not possible no. And I'm not going to add a field for it I'm afraid. You'll have to hack something together yourself.

**Reply by Thulkash** [reply]
Ok, so how could I force the cron job with this
```
# cronjob for Firefly III that changes the target date.
0 3 1 * * /usr/bin/php /var/www/html/artisan firefly-iii:cron --force --date=$(date "+\%Y-\%m-")10
```
whereas
`Cron jobs in Docker
The Docker image does not support cron jobs.` ?

**Reply by JC5** [reply]
So what problem are you trying to solve in the first place?

**Reply by Thulkash** [reply]
Executing future reccuring transactions in a docker-compose environment (3 containers: firefly-core, mysql db and cron)

**Reply by JC5** [reply]
Yes but why?

**Reply by Thulkash** [reply]
I need this because my VISA cards are only debited the first week day of the month "M+1" (delayed debit card) and I refill my account each end of the month "M" like this:
.VISA amount of month "M" (the current one)
.+ all reccuring transactions of month "M+1"
.+ difference to only have 300€ left on my bank account.

I'm using Grisbi since few years for my personal finances. I set its parameters to execute all reccuring transactions of "current day + 25 days".
Thus, each end of the month M, I know how much money I need to transfer on my bank account to cover the month "M+1" (+ 300€ left on my account, just in case)

If I can't execute future recurring transactions, I can't know how much money I need to transfer on my bank account for the next month

**Reply by JC5** [reply, marked as answer]
Mmm, that's complex indeed. It's not impossible to make Firefly III generate future transactions, just not with the tools you're using. A basic bash script that calls `docker exec firefly-iii-app-1 php artisan firefly-iii:cron --date=aaaa-bb-cc` would do the trick already.

**Reply by Thulkash** [reply]
Ok, I'll do that.
Thanks you very much for your time and your help! Much appreciated.

**Comment by JC5** [comment]
> This is pretty confusing.

You can do both. But a static token is easier in Docker.

**Comment by JC5** [comment]
> What is the point?

Google "Docker secrets management"

**Reply by Thulkash** [reply]
Yeah, I know Docker secrets.
I was just pointing that we need to write the secret token in the cron http request in the YAML file (command: sh -c "echo \"0 3 * * * wget -qO- http://app:8080/api/v1/cron/[TOKEN]\" | crontab - && crond -f -L /dev/stdout").

So the secret file is useless in this case.

**Reply by JC5** [reply]
Yes, in this particular case indeed. But there are more ways to run cron jobs :)

**Reply by Thulkash** [reply]
Ok, thanks. I was wondering if there was a better way to do that :)
